core/src/data/properties.cpp

```diff
@@ -104,11 +109,32 @@ void Properties::sort() {
 }
 
 void Properties::add(std::string key, std::string value) {
-    props.emplace_back(std::move(key), Value{std::move(value)});
+
+    auto it = std::lower_bound(props.begin(), props.end(), key,
+                                 [](auto& item, auto& key) {
+                                     return keyComparator(item.key, key);
+                                 });
+
+    if (it == props.end() || it->key != key) {
+        props.emplace_back(std::move(key), Value{std::move(value)});
+    } else {
+        it->value = Value{std::move(value)};
+    }
     sort();
 }
+
 void Properties::add(std::string key, double value) {
-    props.emplace_back(std::move(key), Value{value});
+
+    auto it = std::lower_bound(props.begin(), props.end(), key,
+                                 [](auto& item, auto& key) {
+                                     return keyComparator(item.key, key);
+                                 });
+
+    if (it == props.end() || it->key != key) {
+        props.emplace_back(std::move(key), Value{value});
+    } else {
+        it->value = Value{value};
+    }
```

Comment:
This is good as it also ensures the intended behavior of having unique keys. Could you rename the functions to 'set' ?